types/src/proof/definition.rs

```diff
@@ -470,6 +471,143 @@ impl From<AccumulatorConsistencyProof> for crate::proto::types::AccumulatorConsi
+    /// of the first leaf and all of the leaves in possession.
+    pub fn verify(
+        &self,
+        expected_root_hash: HashValue,
+        first_leaf_index: Option<u64>,
+        leaf_hashes: &[HashValue],
+    ) -> Result<()> {
+        if first_leaf_index.is_none() {
+            ensure!(
+                leaf_hashes.is_empty(),
+                "first_leaf_index indicated empty list while leaf_hashes is not empty.",
+            );
+            ensure!(
+                self.left_siblings.is_empty() && self.right_siblings.is_empty(),
+                "No siblings are needed.",
+            );
+            return Ok(());
+        }
+
+        ensure!(
+            self.left_siblings.len() <= MAX_ACCUMULATOR_PROOF_DEPTH,
+            "Proof has more than {} ({}) left siblings.",
+            MAX_ACCUMULATOR_PROOF_DEPTH,
+            self.left_siblings.len(),
+        );
+        ensure!(
+            self.right_siblings.len() <= MAX_ACCUMULATOR_PROOF_DEPTH,
+            "Proof has more than {} ({}) right siblings.",
+            MAX_ACCUMULATOR_PROOF_DEPTH,
+            self.right_siblings.len(),
+        );
+        ensure!(
+            !leaf_hashes.is_empty(),
+            "leaf_hashes is empty while first_leaf_index indicated non-empty list.",
+        );
+
+        let mut left_sibling_iter = self.left_siblings.iter().rev().peekable();
+        let mut right_sibling_iter = self.right_siblings.iter().rev().peekable();
+
+        let mut first_pos = Position::from_leaf_index(
+            first_leaf_index.expect("first_leaf_index should not be None."),
+        );
+        let mut current_hashes = leaf_hashes.to_vec();
+        let mut parent_hashes = vec![];
+
+        // Keep reducing the list of hashes by combining all the children pairs, until there is
+        // only one hash left.
+        while current_hashes.len() > 1
+            || left_sibling_iter.peek().is_some()
+            || right_sibling_iter.peek().is_some()
+        {
+            let mut leaf_iter = current_hashes.iter();
+            // If the first leaf is a right child, it needs to be combined with a sibling on the
+            // left.
+            if first_pos.is_right_child() {
+                let left_hash = *left_sibling_iter.next().ok_or_else(|| {
+                    format_err!("First child is a right child, but missing sibling on the left.")
+                })?;
+                let right_hash = *leaf_iter.next().expect("The first leaf must exist.");
+                parent_hashes.push(MerkleTreeInternalNode::<H>::new(left_hash, right_hash).hash());
+            }
+
+            // Similarly, if the last leaf is a left child, it needs to be combined with a sibling
+            // on the right. Everything else should form an even number of leaves.
+            for chunk in leaf_iter.as_slice().chunks(2) {
+                let left_hash = chunk[0];
+                let right_hash = if chunk.len() == 2 {
```

Comment:
(I'm not sure, but:) Will it be more clear if we use `chunks_exact()` and `remainder()`?

Reply:
Probably.

```diff
             // Similarly, if the last leaf is a left child, it needs to be combined with a sibling
             // on the right. Everything else should form an even number of leaves.
-            for chunk in leaf_iter.as_slice().chunks(2) {
+            let mut leaf_iter = leaf_iter.as_slice().chunks_exact(2);
+            while let Some(chunk) = leaf_iter.next() {
                 let left_hash = chunk[0];
-                let right_hash = if chunk.len() == 2 {
-                    chunk[1]
-                } else {
-                    *right_sibling_iter.next().ok_or_else(|| {
-                        format_err!("Last child is a left child, but missing sibling on the right.")
-                    })?
-                };
+                let right_hash = chunk[1];
+                parent_hashes.push(MerkleTreeInternalNode::<H>::new(left_hash, right_hash).hash());
+            }
+
+            let remainder = leaf_iter.remainder();
+            assert!(remainder.len() <= 1);
+            if !remainder.is_empty() {
+                let left_hash = remainder[0];
+                let right_hash = *right_sibling_iter.next().ok_or_else(|| {
+                    format_err!("Last child is a left child, but missing sibling on the right.")
+                })?;
                 parent_hashes.push(MerkleTreeInternalNode::<H>::new(left_hash, right_hash).hash());
             }

```